lib/elixir/lib/stream.ex

```diff
@@ -1021,25 +1024,28 @@ defmodule Stream do
     end
   end
 
-  defp do_zip([{fun, fun_acc} | t], acc, callback, list, buffer) do
+  defp do_zip([{fun, fun_acc} | t], acc, callback, yielded_elems, buffer) do
     case fun.({:cont, fun_acc}) do
       {:suspended, [i | fun_acc], fun} ->
-        do_zip(t, acc, callback, [i | list], [{fun, fun_acc} | buffer])
+        do_zip(t, acc, callback, [i | yielded_elems], [{fun, fun_acc} | buffer])
       {_, _} ->
         do_zip_close(:lists.reverse(buffer, t))
         {:done, acc}
     end
   end
 
-  defp do_zip([], acc, callback, list, buffer) do
-    zipped = List.to_tuple(:lists.reverse(list))
+  defp do_zip([], acc, callback, yielded_elems, buffer) do
+    # Here, "list" is the (reversed) list of results of this step of the
+    # enumeration (i.e., if we were zipping three streams, "list" would be a
+    # three-elements list of the elements currently yielded by each stream). We
+    # reverse this to fix the order and convert to tuple since the result of
+    # zipping n streams is a list of n-elements tuples.
+    zipped = List.to_tuple(:lists.reverse(yielded_elems))
     {:next, :lists.reverse(buffer), callback.(zipped, acc)}
   end
 
-  defp do_zip_close([]), do: :ok
-  defp do_zip_close([{fun, acc} | t]) do
-    fun.({:halt, acc})
-    do_zip_close(t)
+  defp do_zip_close(zips) when is_list(zips) do
```

Comment:
Ship it btw!

On Sunday, August 28, 2016, Andrea Leopardi <notifications@github.com>
wrote:

> In lib/elixir/lib/stream.ex
> <https://github.com/elixir-lang/elixir/pull/5169#discussion_r76532691>:
>
> >      {:next, :lists.reverse(buffer), callback.(zipped, acc)}
> >    end
> >
> > -  defp do_zip_close([]), do: :ok
> > -  defp do_zip_close([{fun, acc} | t]) do
> > -    fun.({:halt, acc})
> > -    do_zip_close(t)
> > +  defp do_zip_close(zips) when is_list(zips) do
>
> 👍, updated!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/5169/files/a8bdd32bd7616bd903e3b2385cb978b82cb0c99d#r76532691>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAlbu186cPN9XzEt1eQLQ3dOKqS4x-oks5qkZf6gaJpZM4Ju5Jq>
> .
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*